File: packages/telemetry/typescript/src/sdk/span-filter.ts
import { type Context, isValidSpanId, trace } from "@opentelemetry/api"
import type { ReadableSpan, Span, SpanProcessor } from "@opentelemetry/sdk-trace-node"
import { SCOPE_LATITUDE } from "../constants/scope.ts"

const MAX_TRACKED_SPANS = 2048
const DROPPED_RETENTION_SLACK = 2048

type DroppedSpanEntry = {
  span: ReadableSpan
  recordedParentId: string | undefined
}

type ExportFilterSpanProcessorOptions = {
  blockedInstrumentationScopes?: readonly string[]
}

function parentSpanIdOf(span: ReadableSpan | Span): string | undefined {
  const fromContext = (span as ReadableSpan).parentSpanContext?.spanId
  if (fromContext && isValidSpanId(fromContext)) return fromContext
  const legacy = (span as { parentSpanId?: string }).parentSpanId
  if (legacy && isValidSpanId(legacy)) return legacy
  return undefined
}

function spanIdOf(span: ReadableSpan | Span): string | undefined {
  const spanContext = span.spanContext?.()
  const spanId = spanContext?.spanId
  if (spanId && isValidSpanId(spanId)) return spanId
  return undefined
}

function evictOldestKey(map: Map<string, unknown>): void {
  const oldest = map.keys().next().value
  if (oldest !== undefined) map.delete(oldest)
}

function touchParentTracking(
  parentId: string,
  parentBySpanId: Map<string, string | undefined>,
  droppedBySpanId: Map<string, DroppedSpanEntry>,
  droppedEligibleAt: Map<string, number>,
  dropGeneration: number,
): void {
  if (parentBySpanId.has(parentId)) {
    const recorded = parentBySpanId.get(parentId)
    parentBySpanId.delete(parentId)
    parentBySpanId.set(parentId, recorded)
  }
  const dropped = droppedBySpanId.get(parentId)
  if (dropped) {
    droppedBySpanId.delete(parentId)
    droppedBySpanId.set(parentId, dropped)
    droppedEligibleAt.set(parentId, dropGeneration + DROPPED_RETENTION_SLACK)
  }
}

/** OpenTelemetry GenAI semantic convention attribute prefix. */
const GEN_AI_PREFIX = "gen_ai."

/** Legacy / OpenInference-style LLM attribute prefix. */
const LLM_PREFIX = "llm."

const OPENINFERENCE_KIND = "openinference.span.kind"

/** OTel Python instrumentation scope prefixes for LLM-related instrumentors we support. */
const OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES = [
  "opentelemetry.instrumentation.alephalpha",
  "opentelemetry.instrumentation.anthropic",
  "opentelemetry.instrumentation.bedrock",
  "opentelemetry.instrumentation.cohere",
  "opentelemetry.instrumentation.crewai",
  "opentelemetry.instrumentation.google_generativeai",
  "opentelemetry.instrumentation.groq",
  "opentelemetry.instrumentation.haystack",
  "opentelemetry.instrumentation.langchain",
  "opentelemetry.instrumentation.llamaindex",
  "opentelemetry.instrumentation.mistralai",
  "opentelemetry.instrumentation.ollama",
  "opentelemetry.instrumentation.openai",
  "opentelemetry.instrumentation.replicate",
  "opentelemetry.instrumentation.sagemaker",
  "opentelemetry.instrumentation.together",
  "opentelemetry.instrumentation.transformers",
  "opentelemetry.instrumentation.vertexai",
  "opentelemetry.instrumentation.watsonx",
  "openinference.instrumentation",
] as const

/** Substrings in scope names that indicate LLM / GenAI instrumentation (e.g. Traceloop JS). */
const LLM_SCOPE_SUBSTRINGS = ["openinference", "traceloop", "langsmith", "litellm"] as const

export type SmartFilterOptions = {
  /**
   * When true, all spans are exported (legacy behavior).
   * Default false — only LLM-relevant spans are exported.
   */
  disableSmartFilter?: boolean
  /**
   * When smart filter is on, also export spans for which this returns true
   * (in addition to {@link isDefaultExportSpan}).
   */
  shouldExportSpan?: (span: ReadableSpan) => boolean
  /** Instrumentation scope names to drop (exact match) even if they pass the default predicate. */
  blockedInstrumentationScopes?: string[]
}

/** Input for {@link buildShouldExportSpanFromFields}; allows `undefined` field values for ergonomics. */
export type SmartFilterFieldsInput = {
  disableSmartFilter?: boolean | undefined
  shouldExportSpan?: ((span: ReadableSpan) => boolean) | undefined
  blockedInstrumentationScopes?: string[] | undefined
}

/**
 * Builds the export predicate from loose option fields (`exactOptionalPropertyTypes`-safe call sites).
 */
export function buildShouldExportSpanFromFields(fields: SmartFilterFieldsInput): (span: ReadableSpan) => boolean {
  return buildShouldExportSpan({
    ...(fields.disableSmartFilter !== undefined ? { disableSmartFilter: fields.disableSmartFilter } : {}),
    ...(fields.shouldExportSpan !== undefined ? { shouldExportSpan: fields.shouldExportSpan } : {}),
    ...(fields.blockedInstrumentationScopes !== undefined
      ? { blockedInstrumentationScopes: fields.blockedInstrumentationScopes }
      : {}),
  })
}

function attributeKeys(span: ReadableSpan): string[] {
  const attrs = span.attributes
  if (!attrs || typeof attrs !== "object") return []
  return Object.keys(attrs as Record<string, unknown>)
}

function instrumentationScopeName(span: ReadableSpan): string {
  return span.instrumentationScope?.name ?? ""
}

/** True if the span uses OpenTelemetry GenAI semantic conventions or common LLM attribute namespaces. */
export function isGenAiOrLlmAttributeSpan(span: ReadableSpan): boolean {
  for (const key of attributeKeys(span)) {
    if (key.startsWith(GEN_AI_PREFIX) || key.startsWith(LLM_PREFIX)) return true
    if (key === OPENINFERENCE_KIND || key.startsWith("openinference.")) return true
    // Vercel AI SDK uses ai.* prefix
    if (key.startsWith("ai.")) return true
    // Eve and Flue framework grouping spans carry framework-specific attributes
    if (key.startsWith("eve.") || key.startsWith("flue.")) return true
    // Latitude context attributes
    if (key.startsWith("latitude.")) return true
  }
  return false
}

/** True if the span was created with Latitude's tracer scopes. */
export function isLatitudeInstrumentationSpan(span: ReadableSpan): boolean {
  const name = instrumentationScopeName(span)
  return name === SCOPE_LATITUDE || name.startsWith(`${SCOPE_LATITUDE}.`)
}

function isKnownLlmInstrumentationScope(span: ReadableSpan): boolean {
  const name = instrumentationScopeName(span)
  if (!name) return false
  for (const prefix of OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES) {
    if (name === prefix || name.startsWith(`${prefix}.`)) return true
  }
  const lower = name.toLowerCase()
  for (const part of LLM_SCOPE_SUBSTRINGS) {
    if (lower.includes(part)) return true
  }
  return false
}

/**
 * Default export predicate (smart filter): Latitude scopes, GenAI / LLM attributes,
 * or known LLM instrumentation scopes.
 */
export function isDefaultExportSpan(span: ReadableSpan): boolean {
  if (isLatitudeInstrumentationSpan(span)) return true
  if (isGenAiOrLlmAttributeSpan(span)) return true
  if (isKnownLlmInstrumentationScope(span)) return true
  return false
}

export function buildShouldExportSpan(options: SmartFilterOptions): (span: ReadableSpan) => boolean {
  if (options.disableSmartFilter) return () => true
  const blocked = new Set(options.blockedInstrumentationScopes ?? [])
  const extra = options.shouldExportSpan
  return (span: ReadableSpan) => {
    const scope = instrumentationScopeName(span)
    if (blocked.has(scope)) return false
    if (isDefaultExportSpan(span)) return true
    if (extra?.(span)) return true
    return false
  }
}

/**
 * Drops spans that fail the export predicate; when a span is kept, also exports its ancestors
 * so Latitude receives a connected tree.
 */
export class ExportFilterSpanProcessor implements SpanProcessor {
  private readonly shouldExport: (span: ReadableSpan) => boolean
  private readonly inner: SpanProcessor
  private readonly blockedScopes: ReadonlySet<string>
  private readonly parentBySpanId = new Map<string, string | undefined>()
  private readonly forceExportIds = new Set<string>()
  private readonly droppedBySpanId = new Map<string, DroppedSpanEntry>()
  private readonly droppedEligibleAt = new Map<string, number>()
  private dropGeneration = 0

  constructor(
    shouldExport: (span: ReadableSpan) => boolean,
    inner: SpanProcessor,
    options?: ExportFilterSpanProcessorOptions,
  ) {
    this.shouldExport = shouldExport
    this.inner = inner
    this.blockedScopes = new Set(options?.blockedInstrumentationScopes ?? [])
  }

  onStart(span: Span, parentContext: Context): void {
    const spanId = spanIdOf(span)
    if (spanId) {
      let parentId = parentSpanIdOf(span)
      if (!parentId) {
        const fromCtx = trace.getSpanContext(parentContext)?.spanId
        if (fromCtx && isValidSpanId(fromCtx)) parentId = fromCtx
      }
      if (parentId) {
        touchParentTracking(
          parentId,
          this.parentBySpanId,
          this.droppedBySpanId,
          this.droppedEligibleAt,
          this.dropGeneration,
        )
      }
      if (this.parentBySpanId.size >= MAX_TRACKED_SPANS) evictOldestKey(this.parentBySpanId)
      this.parentBySpanId.set(spanId, parentId)
    }
    this.inner.onStart(span, parentContext)
  }

  onEnd(span: ReadableSpan): void {
    const spanId = spanIdOf(span)
    const recordedParentId = spanId ? this.parentBySpanId.get(spanId) : undefined
    const forced = spanId !== undefined && this.forceExportIds.has(spanId)
    if (spanId) {
      this.forceExportIds.delete(spanId)
      this.parentBySpanId.delete(spanId)
    }

    if (this.isBlocked(span)) {
      this.rememberDropped(span, recordedParentId)
      if (forced) this.flushPromotedAncestors(span, recordedParentId)
      return
    }

    if (!forced && !this.shouldExport(span)) {
      this.rememberDropped(span, recordedParentId)
      return
    }

    if (spanId) this.droppedBySpanId.delete(spanId)
    this.flushPromotedAncestors(span, recordedParentId)
    this.inner.onEnd(span)
  }

  forceFlush(): Promise<void> {
    return this.inner.forceFlush()
  }

  shutdown(): Promise<void> {
    this.parentBySpanId.clear()
    this.forceExportIds.clear()
    this.droppedBySpanId.clear()
    this.droppedEligibleAt.clear()
    return this.inner.shutdown()
  }

  private isBlocked(span: ReadableSpan): boolean {
    return this.blockedScopes.has(instrumentationScopeName(span))
  }

  private rememberDropped(span: ReadableSpan, recordedParentId: string | undefined): void {
    const spanId = spanIdOf(span)
    if (!spanId) return
    this.dropGeneration += 1
    this.droppedEligibleAt.set(spanId, this.dropGeneration + DROPPED_RETENTION_SLACK)
    if (this.droppedBySpanId.size >= MAX_TRACKED_SPANS) {
      for (const key of this.droppedBySpanId.keys()) {
        const eligibleAt = this.droppedEligibleAt.get(key) ?? 0
        if (this.dropGeneration > eligibleAt) {
          this.droppedBySpanId.delete(key)
          this.droppedEligibleAt.delete(key)
          break
        }
      }
    }
    this.droppedBySpanId.set(spanId, { span, recordedParentId })
  }

  private flushPromotedAncestors(span: ReadableSpan, recordedParentId: string | undefined): void {
    for (const ancestor of this.collectPromotedAncestors(span, recordedParentId)) {
      this.inner.onEnd(ancestor)
    }
  }

  private collectPromotedAncestors(span: ReadableSpan, recordedParentId: string | undefined): ReadableSpan[] {
    const toExport: ReadableSpan[] = []
    let parentId = parentSpanIdOf(span) ?? recordedParentId
    const seen = new Set<string>()

    while (parentId && !seen.has(parentId)) {
      seen.add(parentId)

      const dropped = this.droppedBySpanId.get(parentId)
      if (dropped) {
        this.droppedBySpanId.delete(parentId)
        toExport.push(...this.collectPromotedAncestors(dropped.span, dropped.recordedParentId))
        if (!this.isBlocked(dropped.span)) toExport.push(dropped.span)
        parentId = parentSpanIdOf(dropped.span) ?? dropped.recordedParentId
        continue
      }

      if (this.parentBySpanId.has(parentId)) {
        if (this.forceExportIds.size >= MAX_TRACKED_SPANS) {
          const oldest = this.forceExportIds.values().next().value
          if (oldest !== undefined) this.forceExportIds.delete(oldest)
        }
        this.forceExportIds.add(parentId)
        parentId = this.parentBySpanId.get(parentId)
        continue
      }

      break
    }

    return toExport
  }
}

/** Runs optional redaction then the export processor (batch/simple). */
export class RedactThenExportSpanProcessor implements SpanProcessor {
  private readonly redact: SpanProcessor | null
  private readonly exportProcessor: SpanProcessor

  constructor(redact: SpanProcessor | null, exportProcessor: SpanProcessor) {
    this.redact = redact
    this.exportProcessor = exportProcessor
  }

  onStart(span: Span, parentContext: Context): void {
    this.redact?.onStart(span, parentContext)
    this.exportProcessor.onStart(span, parentContext)
  }

  onEnd(span: ReadableSpan): void {
    this.redact?.onEnd(span)
    this.exportProcessor.onEnd(span)
  }

  forceFlush(): Promise<void> {
    return this.exportProcessor.forceFlush()
  }

  shutdown(): Promise<void> {
    return this.exportProcessor.shutdown()
  }
}

File: packages/telemetry/python/src/latitude_telemetry/telemetry/span_filter.py
"""
Smart export filter: only LLM-relevant spans are sent to Latitude by default.
"""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass

from opentelemetry.context import Context
from opentelemetry.sdk.trace import ReadableSpan, Span, SpanProcessor
from opentelemetry.trace import INVALID_SPAN_ID, format_span_id, get_current_span

from latitude_telemetry.constants import SCOPE_LATITUDE

GEN_AI_PREFIX = "gen_ai."
LLM_PREFIX = "llm."
OPENINFERENCE_KIND = "openinference.span.kind"

# Cap for ended-but-dropped spans held so a later child can still promote them.
_MAX_DROPPED_SPAN_BUFFER = 2048

OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES: tuple[str, ...] = (
    "opentelemetry.instrumentation.alephalpha",
    "opentelemetry.instrumentation.anthropic",
    "opentelemetry.instrumentation.bedrock",
    "opentelemetry.instrumentation.cohere",
    "opentelemetry.instrumentation.crewai",
    "opentelemetry.instrumentation.google_generativeai",
    "opentelemetry.instrumentation.groq",
    "opentelemetry.instrumentation.haystack",
    "opentelemetry.instrumentation.langchain",
    "opentelemetry.instrumentation.llamaindex",
    "opentelemetry.instrumentation.mistralai",
    "opentelemetry.instrumentation.ollama",
    "opentelemetry.instrumentation.openai",
    "opentelemetry.instrumentation.replicate",
    "opentelemetry.instrumentation.sagemaker",
    "opentelemetry.instrumentation.together",
    "opentelemetry.instrumentation.transformers",
    "opentelemetry.instrumentation.vertexai",
    "opentelemetry.instrumentation.watsonx",
    "openinference.instrumentation",
)

LLM_SCOPE_SUBSTRINGS: tuple[str, ...] = ("openinference", "traceloop", "langsmith", "litellm")


def _attribute_keys(span: ReadableSpan) -> list[str]:
    attrs = getattr(span, "attributes", None)
    if not attrs:
        return []
    return list(attrs.keys())


def _instrumentation_scope_name(span: ReadableSpan) -> str:
    scope = getattr(span, "instrumentation_scope", None)
    if scope is None:
        return ""
    return getattr(scope, "name", "") or ""


def _span_id(span: ReadableSpan | Span) -> str:
    return format_span_id(span.get_span_context().span_id)


def _parent_span_id(span: ReadableSpan | Span) -> str | None:
    parent = getattr(span, "parent", None)
    if parent is None:
        return None
    span_id = getattr(parent, "span_id", None)
    if span_id is None or span_id == INVALID_SPAN_ID:
        return None
    return format_span_id(span_id)


def is_gen_ai_or_llm_attribute_span(span: ReadableSpan) -> bool:
    for key in _attribute_keys(span):
        if key.startswith(GEN_AI_PREFIX) or key.startswith(LLM_PREFIX):
            return True
        if key == OPENINFERENCE_KIND or key.startswith("openinference."):
            return True
        # Vercel AI SDK uses ai.* prefix
        if key.startswith("ai."):
            return True
        if key.startswith("latitude."):
            return True
    return False


def is_latitude_instrumentation_span(span: ReadableSpan) -> bool:
    name = _instrumentation_scope_name(span)
    return name == SCOPE_LATITUDE or name.startswith(f"{SCOPE_LATITUDE}.")


def _is_known_llm_instrumentation_scope(span: ReadableSpan) -> bool:
    name = _instrumentation_scope_name(span)
    if not name:
        return False
    for prefix in OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES:
        if name == prefix or name.startswith(f"{prefix}."):
            return True
    lower = name.lower()
    return any(part in lower for part in LLM_SCOPE_SUBSTRINGS)


def is_default_export_span(span: ReadableSpan) -> bool:
    if is_latitude_instrumentation_span(span):
        return True
    if is_gen_ai_or_llm_attribute_span(span):
        return True
    if _is_known_llm_instrumentation_scope(span):
        return True
    return False


@dataclass
class SmartFilterOptions:
    """Options for the default export predicate."""

    disable_smart_filter: bool = False
    should_export_span: Callable[[ReadableSpan], bool] | None = None
    blocked_instrumentation_scopes: tuple[str, ...] = ()


def build_should_export_span(
    *,
    disable_smart_filter: bool = False,
    should_export_span: Callable[[ReadableSpan], bool] | None = None,
    blocked_instrumentation_scopes: Sequence[str] | None = None,
) -> Callable[[ReadableSpan], bool]:
    if disable_smart_filter:
        return lambda _span: True
    blocked = set(blocked_instrumentation_scopes or ())
    extra = should_export_span

    def should_export(span: ReadableSpan) -> bool:
        scope = _instrumentation_scope_name(span)
        if scope in blocked:
            return False
        if is_default_export_span(span):
            return True
        if extra is not None and extra(span):
            return True
        return False

    return should_export


class ExportFilterSpanProcessor(SpanProcessor):
    """Drops filtered spans; when a span is kept, also exports its ancestors."""

    def __init__(
        self,
        should_export: Callable[[ReadableSpan], bool],
        inner: SpanProcessor,
    ) -> None:
        self._should_export = should_export
        self._inner = inner
        self._parent_by_span_id: dict[str, str | None] = {}
        self._force_export_ids: set[str] = set()
        self._dropped_by_span_id: dict[str, ReadableSpan] = {}

    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        span_id = _span_id(span)
        parent_id = _parent_span_id(span)
        if parent_id is None and parent_context is not None:
            parent_ctx = get_current_span(parent_context).get_span_context()
            if parent_ctx.is_valid and parent_ctx.span_id != INVALID_SPAN_ID:
                parent_id = format_span_id(parent_ctx.span_id)
        self._parent_by_span_id[span_id] = parent_id
        self._inner.on_start(span, parent_context)

    def on_end(self, span: ReadableSpan) -> None:
        span_id = _span_id(span)
        recorded_parent_id = self._parent_by_span_id.pop(span_id, None)
        forced = span_id in self._force_export_ids
        self._force_export_ids.discard(span_id)

        if not forced and not self._should_export(span):
            self._remember_dropped(span)
            return

        self._dropped_by_span_id.pop(span_id, None)
        self._promote_ancestors(span, recorded_parent_id)
        self._inner.on_end(span)

    def shutdown(self) -> None:
        self._parent_by_span_id.clear()
        self._force_export_ids.clear()
        self._dropped_by_span_id.clear()
        self._inner.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return self._inner.force_flush(timeout_millis)

    def _remember_dropped(self, span: ReadableSpan) -> None:
        if len(self._dropped_by_span_id) >= _MAX_DROPPED_SPAN_BUFFER:
            oldest = next(iter(self._dropped_by_span_id))
            del self._dropped_by_span_id[oldest]
        self._dropped_by_span_id[_span_id(span)] = span

    def _promote_ancestors(self, span: ReadableSpan, recorded_parent_id: str | None = None) -> None:
        parent_id = _parent_span_id(span) or recorded_parent_id
        seen: set[str] = set()

        while parent_id and parent_id not in seen:
            seen.add(parent_id)

            dropped = self._dropped_by_span_id.pop(parent_id, None)
            if dropped is not None:
                self._promote_ancestors(dropped)
                self._inner.on_end(dropped)
                parent_id = _parent_span_id(dropped)
                continue

            if parent_id in self._parent_by_span_id:
                self._force_export_ids.add(parent_id)
                parent_id = self._parent_by_span_id.get(parent_id)
                continue

            break


class RedactThenExportSpanProcessor(SpanProcessor):
    """Runs optional redaction then the batch/simple export processor."""

    def __init__(
        self,
        redact: SpanProcessor | None,
        export_processor: SpanProcessor,
    ) -> None:
        self._redact = redact
        self._export_processor = export_processor

    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        if self._redact is not None:
            self._redact.on_start(span, parent_context)
        self._export_processor.on_start(span, parent_context)

    def on_end(self, span: ReadableSpan) -> None:
        if self._redact is not None:
            self._redact.on_end(span)
        self._export_processor.on_end(span)

    def shutdown(self) -> None:
        self._export_processor.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return self._export_processor.force_flush(timeout_millis)
